kernel/src/main/java/org/kframework/parser/concrete2kore/disambiguation/KAppToTermConsVisitor.java

```diff
@@ -43,57 +51,68 @@ public KAppToTermConsVisitor(Module mod) {
-            // prefer KList if they exist, otherwise return t
-            List<PStack<Term>> klists = amb.items().stream()
-                    .filter(x -> x instanceof TermCons
-                            && ((TermCons) x).production().klabel().isDefined()
-                            && ((TermCons) x).production().klabel().get().name().equals("#KList"))
-                    .map(KAppToTermConsVisitor::flattenKList)
-                    .sorted((o1, o2) -> o2.size() - o1.size()).collect(Collectors.toList());
-
-            // expecting the first ambiguity branch to have the highest amount of elements
-            assert klists.size() <= 1 || klists.get(0).size() != klists.get(1).size():
-                    KAppToTermConsVisitor.class + ":" + " unexpected ambiguity pattern found under KApp " + amb;
-            if (klists.size() != 0)
-                return klists.get(0);
+            assert false : KAppToTermConsVisitor.class + " expected all ambiguities to already be pushed to the top:\n" +
+                    "   Source: " + ((Ambiguity) t).items().iterator().next().source().orElse(null) + "\n" +
+                    "   Location: " + ((Ambiguity) t).items().iterator().next().location().orElse(null);
         } else if (t instanceof TermCons) {
             TermCons tc = (TermCons) t;
-            if (tc.production().klabel().isDefined() && tc.production().klabel().get().name().equals("#KList"))
-                return flattenKList(tc.items().get(1)).plusAll(flattenKList(tc.items().get(0)));
-            else if (tc.production().klabel().isDefined() && tc.production().klabel().get().name().equals("#EmptyKList"))
+            if (tc.production().klabel().isDefined() && tc.production().klabel().get().equals(KLabels.KLIST))
+                return flattenKList(tc.get(0)).plusAll(flattenKList(tc.get(1)));
+            else if (tc.production().klabel().isDefined() && tc.production().klabel().get().equals(KLabels.EMPTYKLIST))
                 return ConsPStack.empty();
         }
         return ConsPStack.singleton(t);
     }
+
+    // push ambiguities top so we can get easy access to KList
+    private static class PushTopAmbiguityUp2 extends SafeTransformer {
+        @Override
+        public Term apply(TermCons tc) {
+            if (tc.production().klabel().isDefined() && tc.production().klabel().get().head().equals(KLabels.KLIST)) {
+                Term v0 = super.apply(tc.get(0));
+                Term v1 = super.apply(tc.get(1));
+                Set<Term> t0 = v0 instanceof Ambiguity ? ((Ambiguity) v0).items() : Sets.newHashSet(v0);
+                Set<Term> t1 = v1 instanceof Ambiguity ? ((Ambiguity) v1).items() : Sets.newHashSet(v1);
+                Set<Term> rez = Sets.newHashSet();
+                for (Term t00 : t0)
+                    for (Term t11 : t1)
+                        rez.add(TermCons.apply(ConsPStack.singleton(t00).plus(t11), tc.production(), tc.location(), tc.source()));
```

Comment:
Pretty sure this is reversed from the order you want.

Reply:
This is wrong, I misunderstood that plus does a prepend in constant time rather than an append.